### ISSUE TITLE
How to approach loading of .scss files

### ISSUE THREAD

**Issue by patroza** [aurelia/webpack-plugin#14]
It seems the aurelia-templating-resources has specific support for loading .css files,
i've found that I can make scss files work by adding:

```
  viewEngine.addResourcePlugin('.scss', {
    'fetch': function fetch(address) {
      var _ref;

      return _ref = {}, _ref[address] = (0, _cssResource._createCSSResource)(address), _ref;
    }
  });
```

while using this loader configuration:

```
{ test: /\.scss$/, loaders: ["raw", "sass"] },
```

**Comment by wesleyb-io** [comment]
These were my steps to enable sass with Aurelia and Webpack.

0. Use [aurelia/skeleton-navigation/skeleton-es2016-webpack@1.0.0-beta.1.1.6](https://github.com/aurelia/skeleton-navigation/blob/1.0.0-beta.1.1.6/skeleton-es2016-webpack)
0. To [webpack.config.js](https://github.com/aurelia/skeleton-navigation/blob/1.0.0-beta.1.1.6/skeleton-es2016-webpack/webpack.config.js), add `{ test: /\.scss$/, loader: 'style-loader!css-loader!sass-loader' },`.
    - I placed the line underneath `{ test: /\.css?$/, loader: 'style!css' },`.
    - I thought I had done this via [`sass-loader` instructions](https://github.com/jtangelder/sass-loader#apply-via-webpack-config), but my syntax differs.

**Comment by sickboy** [comment]
@wesleyb-io thanks but that would not work from within .html templates <require elements

**Comment by wesleyb-io** [comment]
Ah, yes. After finally getting time to break my SCSS into modules, I've just realized this.

When I searched the [aurelia-templating-resources](https://github.com/aurelia/templating-resources) for `viewEngine.addResourcePlugin`, I found several results. Did you add your code to just the file relevant to your project, or have you found the one file that (I assume) populates all the others?

**Comment by mirronelli** [comment]
you can load scss from html templates using require element as this:
_<template>
	<require from="./app.scss"></require>
	helllo from aurelia
</template>_
when having scss configured as bellow
_module: {
		loaders: [	{ test: /\.scss$/, loader: 'style!css!sass' }	]
	},_

The tricky part is to have the paths right. Webpack needs the "./" or "../" prefix to resolve it as a relative link, otherwise it thinks it is a node_module and does not work. It took me a few hours to fight this one.

**Comment by Vheissu** [comment]
I think this issue should be closed @EisenbergEffect 

This isn't really related to the Webpack plugin at all. It is a loader question, which is something specific to Webpack configuration and not the plugin. The plugin merely is responsible for mapping dependencies used within an Aurelia application or Node modules folder, it is type agnostic in that this plugin doesn't add in support for different file types like Sass.

**Comment by sickboy** [comment]
@EisenbergEffect @Vheissu 
I still can't use scss files with

<require from="./some.scss"></require>
 in my views.

The aurelia-webpack-plugin has this:
~~~
case ".scss":
  output += '!!sass!';
~~~

The result is:
 Error: Module parse failed: ...some.scss Unexpected token (1:10)(…)

Next step we adjust the webpack-plugin:
output += '!!css!!sass!';

now we get:
 Error: Cannot find module './some.scss.html'.
(note the .html)

Next step we add to the aurelia-templating-resources:

~~~
viewEngine.addResourcePlugin('.scss', {
  'fetch': function fetch(address) {
    var _ref;

    return _ref = {}, _ref[address] = (0, _cssResource._createCSSResource)(address), _ref;
  }
});
~~~

And now we have a winner.

**Comment by niieani** [comment]
Yes, we are aware of the fact that loading `.scss` and other custom files is not yet supported by Aurelia.

**Comment by niieani** [comment]
I've fixed the loader code, however `addResourcePlugin` is still your responsibility. @EisenbergEffect any ideas how we can support this globally?

**Comment by EisenbergEffect** [comment]
What is the actual request?

**Comment by niieani** [comment]
@EisenbergEffect The question is: how might we support loading custom CSS files, like `.scss` or `.less` inside of View templates. Currently, the workaround is to manually add a resource plugin for the required extension to the ViewEngine: `viewEngine.addResourcePlugin('.scss', ...)` and make it use the standard `.css` resource plugin. Since Webpack handles compilation of these files, it works transparently. 
The question is whether we should add these other known CSS extension to be handled by default, make a small plugin which does it for people, or something else?

**Comment by EisenbergEffect** [comment]
I'm not sure if we should add all of them by default.

We can make a helper method though...to "map" various extensions to css, essentially so that you can accomplish the above without the unpleasant code writing.

Thoughts?

**Comment by niieani** [comment]
I'm open to any option. How would you see the helper method implemented and where should it be called? During aurelia-framework configuration?

**Comment by EisenbergEffect** [comment]
Yes, it would be called during the framework configuration. I need to think about what that api would look like. It's going to reside on the view engine, since that's what handles the plugin/extension mapping. Maybe we just need a simple "alias" api for file extensions.

**Comment by niieani** [comment]
`alias` seems like a good idea!

**Comment by sickboy** [comment]
alias which can map many extensons to one would be great, so we can map .scss, .sass, .less, etc all to .css.
And who knows, maybe even .haml to .html, and such :) (oh my, the old memories of RoR are coming back)

**Comment by heruan** [comment]
So, in the meantime where and what should one do to enable the `ViewEngine` to load `*.scss` files?

**Comment by Saeris** [comment]
I'd also like to know how to do this! My project's been on hold because I'm waiting on aliasing support.

**Comment by sickboy** [comment]
Here's my fork of the latest with scss support https://github.com/sickboy/templating-resources/commit/bdc0f1309109157c233ca75718abdf556dd7dfbc

**Comment by wichert** [comment]
One thing I am noticing is that it does not seem to be possible to add SCSS support without forking templating-resource because `_createCSSResource` is not accessible. Would it make sense to expose that so adding support for sass/less/etc. can easily be done without having to hardcode all of them in templating-resource?

**Comment by Saeris** [comment]
Yeah I'd much prefer to not have to modify templating-resource if I can, since that would only be a temporary solution at best. Thank you @sickboy for sharing your fork though so we can see how that's done.

**Comment by wichert** [comment]
I added aurelia/templating-resources#241 asking for `_createCSSResource` to become an official API.

**Comment by alpox** [comment]
Any updates on this?

**Comment by xiwc** [comment]
waiting updates...

**Comment by Thanood** [comment]
There's a [config-sass](https://github.com/easy-webpack/config-sass) plugin for easy-webpack..
I don't have any good webpack experience but I think this creates an scss loader for webpack.

**Comment by niieani** [comment]
@Thanood you still need to configure Aurelia if you want to use SASS inside View templates.

**Comment by lares83** [comment]
@niieani can you write how can we configure Aurelia to use SASS inside View templates?
what I also noticed is that `config-sass` with this configuration `require('@easy-webpack/config-sass')({ filename: 'app.css', allChunks: false, sourceMap: false })`,  puts all scss files into one `app.css` even without importing them in JS  files, can you explain me why or maybe this is a bug?

**Comment by Ixonal** [comment]
I tried putting the resource plugin suggested by @sickboy (had to reference the css-resource module directly through node_modules, so not a great solution), but it's saying "BindingLanguage must implement inspectTextContent()" when it tries to load the app. Anyone else get that? Here's what I'm doing...

``` javascript

//in main.ts, doing framework configuration
import { _createCSSResource } from "../node_modules/aurelia-templating-resources/dist/es2015/css-resource";

//...then later

  (aurelia.container.get(ViewEngine) as ViewEngine).addResourcePlugin(".scss", {
    fetch(address: string) {
      var _ref;

      return _ref = {}, _ref[address] = (0, _createCSSResource)(address), _ref;
    }
  });

```

**Comment by Saeris** [comment]
Instead of using <require> statements in the view to load .SCSS files, I'm now using import in the view model. This strategy requires no modification to Aurelia Templating Resources, and instead of producing an embedded stylesheet, imported .scss gets bundled together as a partial in app.css. For now I believe this is the best approach for loading Sass at the component level. Examples can be found in my project here:

https://github.com/Saeris/Compendium

**Comment by niieani** [comment]
@Saeris this approach works if you are okay with merging and bundling the CSS together, i.e. you can ensure there are no naming / scoping conflicts between CSS for various components.

**Comment by Saeris** [comment]
@niieani True. My reason for doing things this way is for code organization; I like to keep the CSS for each component in it's own file along with the view/view model to make editing easier.

Unless you're applying styles directly to elements or using Shadow DOM, there is always going to be some bleed-over with CSS because that's the nature of the language. There are scoping techniques to mitigate this, and the one I use is a top-level class or id under which I nest all the CSS for a particular component, which adds a layer of specificity.

My solution is not the best use-case for everyone, but I thought I'd share it here to help those who want to include Sass with individual components rather than managing Sass imports manually.

**Comment by jods4** [comment]
In addition to `.css`, I added support for `.less`, `.sass` and `.scss` in `aurelia-templating-resources@1.4`.
https://github.com/aurelia/templating-resources/commit/901cfd8

You'll need the proper loaders, for example if you want to `<require from="style.less">` you'll need `css-loader!less-loader` in your Webpack config.

**Comment by MJLHThomassen** [comment]
@jods4 Ive upgraded my aurelia packages to use aurealia-templating-resources version 1.4.0 which contains your addition however i'm still getting errors loading scscc file using the <require> tag. Ive set up my webpack to use the scss loader. Is there somethign else im missing / some aurelia thing i need to configure?

**Comment by jods4** [comment]
You need:
- `aurelia-templating-resources@1.4.0`
- The new Aurelia plugins for webpack (2.0 RC)
- A correct webpack config: scss loader, no style-loader.

That should work, at least I tested LESS.
What error do you get?

**Comment by MaartenThomassen** [comment]
@jods4 

The 3rd point (no style-loader) did it for me, thanks!

**Comment by MJLHThomassen** [comment]
@jods4 

The third point (no style-loader) did it for me, thanks!

**Comment by MrTango** [comment]
i used the aurelia cli with the webpack config, now I'm struggeling adding the sass-loader, do you have an working example somewhere?

I always get:

```
ERROR in ./node_modules/materialize-css/sass/materialize.scss
Module parse failed: Unexpected character '@' (1:0)
You may need an appropriate loader to handle this file type.
| @charset "UTF-8";
|
```

**Comment by Alexander-Taran** [comment]
@MrTango 
```
      {
        test: /\.scss$/,
        use: ['style-loader', 'css-loader', 'sass-loader'],
        issuer: /\.[tj]s$/i
      },
      {
        test: /\.scss$/,
        use: ['css-loader', 'sass-loader'],
        issuer: /\.html?$/i
      },
```
